### DISCUSSION TITLE
Using SCIP solver in primal task

### DISCUSSION THREAD

**Discussion by kushaltirumala** [ds4dm/ml4co-competition#21] [Q&A] [answered]
In the primal task environment, the switch SCIP heuristics are turned off. Does this mean we cannot use SCIP's `model.optimize()​` at all in our agent's Policy `__call__`​ function definition? Or, are we allowed to turn on SCIP heuristics for a different `model`​ that is not the original model​ passed to the `extract​` function?

**Comment by gasse** [comment, marked as answer]
Hi @kushaltirumala ,

You are not allowed to use `model.optimize()​` in your agent, as this would change the state of the model. Your agent is only allowed to extract information from SCIP (within your observation function), and then return an action (variable assignments in the primal task).

Note that you can, however, make a copy of the model using PySCIPOpt as follows:
```
m = model.as_pyscipopt()
m_copy = pyscipopt.Model(sourceModel=m, globalcopy=True)
```

This copy is a perfectly valid observation. You can then use this copy in whichever way you want within your agent, as this will not change the state of the original model, i.e., the inner state of the environment.

PS: note that making such a copy of the model can be costly. Maybe you want to do it only once the first time your primal agent is called.

Hope this helps,

Best,
Maxime
